src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer.UnitTests/FileBasedProgramsWorkspaceTests.cs

```diff
@@ -70,4 +70,65 @@ private protected override Workspace GetHostWorkspace(TestLspServer testLspServe
+        using var tempRoot = new TempRoot();
+        var newDocumentFilePath = Path.Combine(tempRoot.CreateDirectory().Path, "Program.cs");
+        File.WriteAllText(newDocumentFilePath, markup);
+
+        var newDocumentUri = ProtocolConversions.CreateAbsoluteDocumentUri(newDocumentFilePath);
+
+        // Ensure LSP has the text
+        await testLspServer.OpenDocumentAsync(newDocumentUri, "Console.WriteLine();");
+
+        // The first call should return a document in the misc files workspace.
+        var (_, miscDocument) = await GetLspWorkspaceAndDocumentAsync(newDocumentUri, testLspServer).ConfigureAwait(false);
+        Assert.NotNull(miscDocument);
+        Assert.True(await testLspServer.GetManagerAccessor().IsMiscellaneousFilesDocumentAsync(miscDocument));
+        Assert.Equal(WorkspaceKind.MiscellaneousFiles, miscDocument.Project.Solution.Workspace.Kind);
+
+        // Wait for the project system to load the file as a FBP
+        await testLspServer.TestWorkspace.GetService<AsynchronousOperationListenerProvider>().GetWaiter(FeatureAttribute.Workspace).ExpeditedWaitAsync();
+
+        // The second call should return an FBP document in the host workspace.
+        var (documentWorkspace, document) = await GetLspWorkspaceAndDocumentAsync(newDocumentUri, testLspServer).ConfigureAwait(false);
+        Assert.NotNull(document);
+        Assert.NotNull(documentWorkspace);
+        Assert.Equal(WorkspaceKind.Host, documentWorkspace.Kind);
+        // FBP project path is just the document file path.
+        Assert.Equal(newDocumentFilePath, document.Project.FilePath);
+
+        // Now, upadate the host workspace to also include the document (simulating it getting added by the project system later on).
+        var workspaceFactory = testLspServer.TestWorkspace.ExportProvider.GetExportedValue<LanguageServerWorkspaceFactory>();
+        var projectName = "MainSolutionProject";
+        var project = await workspaceFactory.HostProjectFactory.CreateAndAddToWorkspaceAsync(
+            Guid.NewGuid().ToString(),
+            LanguageNames.CSharp,
+            new ProjectSystemProjectCreationInfo { AssemblyName = projectName },
+            workspaceFactory.ProjectSystemHostInfo);
+
+        project.AddSourceFile(newDocumentFilePath);
+
+        // Wait for the project system to load the host project
+        await testLspServer.TestWorkspace.GetService<AsynchronousOperationListenerProvider>().GetWaiter(FeatureAttribute.Workspace).ExpeditedWaitAsync();
+
+        // LSP should return the document from the main solution instead of the FBP.
+        (documentWorkspace, document) = await GetLspWorkspaceAndDocumentAsync(newDocumentUri, testLspServer).ConfigureAwait(false);
+        Assert.NotNull(document);
+        Assert.NotNull(documentWorkspace);
+        Assert.Equal(WorkspaceKind.Host, documentWorkspace.Kind);
+        Assert.Equal(document.Project.AssemblyName, projectName);
```

Comment:
It seems like this should be `Assert.Equal(projectName, document.Project.AssemblyName)` to reflect that `projectName` is the "expected" value